src/datasets/utils/hub.py

```diff
@@ -0,0 +1,18 @@
+from typing import Optional
+from urllib.parse import quote
+
+import huggingface_hub.file_download
+
+from .. import config
+from .py_utils import temporary_assignment
+
+
+def hf_hub_url(repo_id: str, path: str, revision: Optional[str] = None) -> str:
+    revision = revision or config.HUB_DEFAULT_VERSION
+    with temporary_assignment(
+        huggingface_hub.file_download,
+        "HUGGINGFACE_CO_URL_TEMPLATE",
+        config.HUB_DATASETS_URL.replace("{path}", "{filename}"),
+    ):
```

Comment:
cc @Wauplin si it how you'd expect `datasets` to overwrite `HUGGINGFACE_CO_URL_TEMPLATE` ?

Reply:
Wow, what is the goal here ? Is it just to avoid passing `repo_type="dataset"` ?

For what I understand, `HUB_DATASETS_URL` from `datasets` and `HUGGINGFACE_CO_URL_TEMPLATE` from `huggingface_hub` are very similar, isn't it ?

```py
HUGGINGFACE_CO_URL_TEMPLATE = ENDPOINT + "/{repo_id}/resolve/{revision}/{filename}"
HUB_DATASETS_URL = HF_ENDPOINT + "/datasets/{repo_id}/resolve/{revision}/{path}"
```

Reply:
I think the goal is to overwrite the hfh one with the `datasets` one, but idk if we should

Reply:
I would really avoid to as the `"/{repo_id}/resolve/{revision}/{filename}"` part is not specific to `datasets`.
If it is a matter of customizing the endpoint, both are defined as `HF_ENDPOINT = os.environ.get("HF_ENDPOINT", "https://huggingface.co")` so it doesn't add a difference with existing `hfh` behavior.

Reply:
The thing is that `datasets` users might have override the variable `config.HUB_DATASETS_URL`.

If we don't pass this to `hfh`, this is a breaking change and we should warn `datasets` users that their setting `config.HUB_DATASETS_URL` is ignored.

Indeed I struggled to have this working:
- in `datasets` we import the module `config` each time we need to use a variable: then we can modify afterwards one of the variable in it and that will be updated wherever the `config` module is used; we support dynamic update of the config
- in `hfh` only the config constant is imported wherever it is used and this creates a "copy" of it: whenever we modify afterwards the variable value in `huggingface_hub.constants` has no effect on `huggingface_hub.file_download` imported constants: they were already "copied"

In relation with the ENDPOINT, we set it also dynamically in our CI, using fixtures. However, it is useless to modify it dynamically in `hfh`: once the repo URL is defined from the "old" ENDPOINT, changing a new value for the ENDPOINT will not have any effect in the repo URL: this is set only once at import time.

Reply:
The goal here was to avoid the breaking change of ignoring the `config.HUB_DATASETS_URL` setting.

Reply:
Also, while tweaking this, I discovered our CI has some tests that request the production ENDPOINT: so the CI ENDPOINT cannot be set globally for all tests: https://github.com/huggingface/datasets/actions/runs/3386458817/jobs/5625907924
```
TokenizersDumpTest.test_hash_tokenizer

>           raise HTTPError(http_error_msg, response=self)
E           requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://hub-ci.huggingface.co/bert-base-uncased/resolve/main/tokenizer_config.json
```

Reply:
I imagine you need this, @albertvillanova : https://github.com/huggingface/huggingface_hub/issues/1082

Reply:
Thank you @albertvillanova for the broader context. Since how long do you have this `HUB_DATASETS_URL` env variable ? It might be worth to deprecate it if it's only used to overwrite `HUGGINGFACE_CO_URL_TEMPLATE` (if it has a larger scope, maybe it's good to keep if).

> I imagine you need this, @albertvillanova : https://github.com/huggingface/huggingface_hub/issues/1082

Agree that this should be the clean solution to do it :)

Reply:
> in hfh only the config constant is imported wherever it is used and this creates a "copy" of it:

Totally get your point here. Might be worth opening an issue in `hfh` to fix that once and for all :)

Reply:
@Wauplin that fix would make testing much easier... :hugs: